### PR TITLE
Update github action to upload test result files

### PR DESCRIPTION
**Pull request by jingwei99**

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -73,6 +73,18 @@ jobs:
           path: zipline/src/jvmMain/resources/*
           if-no-files-found: error
 
+      - uses: actions/upload-artifact@v3
```
Comment (jingwei99) [inline .github/workflows/build.yaml:70]: notes: delete L70-L75

**Comment by jingwei99** [comment]
it works, github run: https://github.com/cashapp/zipline/actions/runs/7227325146

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -72,6 +72,13 @@ jobs:
           name: jvm-native-libraries
           path: zipline/src/jvmMain/resources/*
           if-no-files-found: error
+      
+      - uses: actions/upload-artifact@v4
```
Comment (jingwei99) [inline .github/workflows/build.yaml]: this seems to [create artifacts with the same name for different flows](https://github.com/cashapp/zipline/actions/runs/7227684902/job/19695773908?pr=1193#step:9:15), looking

Comment (jingwei99) [reply .github/workflows/build.yaml]: update: figured out, added matrix suffixes, and they work well:  https://github.com/cashapp/zipline/actions/runs/7227981446
Screenshot 2023-12-15 at 3 44 39 PM

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -69,10 +69,17 @@ jobs:
 
       - uses: actions/upload-artifact@v4
         with:
-          name: jvm-native-libraries
+          name: jvm-native-libraries-${{ matrix.os }}-${{ matrix.arch }}-${{ matrix.cmake-arch }}
```
Comment (JakeWharton) [inline .github/workflows/build.yaml]: This should be reverted. We want them to merge together.

**Review by JakeWharton — APPROVED** [review]

**Review by swankjesse — APPROVED** [review]

File: .github/workflows/build.yaml (hunk 3 above)
Comment (jingwei99) [reply .github/workflows/build.yaml]: it's [failing](https://github.com/cashapp/zipline/actions/runs/7228285030/job/19697454944#step:9:15) with the [same reason](https://github.com/cashapp/zipline/pull/1200#discussion_r1428571699) after revert, I found [this](https://github.com/actions/upload-artifact?tab=readme-ov-file#not-uploading-to-the-same-artifact) from the `upload-artifact` documentation about `v4`:

Screenshot 2023-12-15 at 4 44 35 PM

To keep the merge together behavior, I'll revert the job uploading `jvm-native-libraries` to use `upload-artifact@v3` .

Comment (JakeWharton) [reply .github/workflows/build.yaml]: We can still change them to be unique, but it means the consuming job will need to download them all and merge locally.

Comment (jingwei99) [reply .github/workflows/build.yaml]: I'll keep it to V3 for this PR to get the uploading test files part in, it sounds moving forward: 

1) this should be updated to V4 with unique names 
2) update the consuming job to download and merge (where do we have the consuming jobs right now?)

Comment (JakeWharton) [reply .github/workflows/build.yaml]: Consuming job is probably named publish. Whichever has the download action for this folder.

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -67,12 +67,20 @@ jobs:
...
           path: zipline/src/jvmMain/resources/*
           if-no-files-found: error
 
+      - uses: actions/upload-artifact@v4
+        if: ${{ always() }}
+        with:
+          name: test-report-${{ matrix.os }}-${{ matrix.arch }}-${{ matrix.cmake-arch }}
+          path: '**/build/reports/tests/**'
+          retention-days: 1
+          compression-level: 9
```
Comment (jingwei99) [inline .github/workflows/build.yaml]: now it's failing with: 

```
Cause 15: org.gradle.api.GradleException: Could not write XML test results for app.cash.zipline.loader.internal.fetcher.HttpFetcherTest to file /Users/runner/work/zipline/zipline/zipline-loader/build/test-results/macosX64Test/TEST-app.cash.zipline.loader.internal.fetcher.HttpFetcherTest.xml.
	at org.gradle.api.internal.tasks.testing.junit.result.Binary2JUnitXmlReportGenerator$JUnitXmlReportFileGenerator.run(Binary2JUnitXmlReportGenerator.java:125)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$QueueWorker.execute(DefaultBuildOperationExecutor.java:238)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.runOperation(DefaultBuildOperationQueue.java:272)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.doRunBatch(DefaultBuildOperationQueue.java:253)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.lambda$runBatch$0(DefaultBuildOperationQueue.java:238)
	at org.gradle.internal.resources.AbstractResourceLockRegistry.whileDisallowingLockChanges(AbstractResourceLockRegistry.java:50)
	at org.gradle.internal.work.DefaultWorkerLeaseService.whileDisallowingProjectLockChanges(DefaultWorkerLeaseService.java:236)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.lambda$runBatch$1(DefaultBuildOperationQueue.java:238)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:264)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:128)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.runBatch(DefaultBuildOperationQueue.java:224)
	at org.gradle.internal.operations.DefaultBuildOperationQueue$WorkerRunnable.run(DefaultBuildOperationQueue.java:192)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1623)
Caused by: com.esotericsoftware.kryo.KryoException: Buffer underflow.
```

added compression level to see if it helps.

ideally this step can be considered as optional: failing this step doesn't fail the whole job. I did a time-boxed search and didn't find a way to achieve this with github action (`continue-on-error` will still give a ❌ for the job)

**Pull request merged**